Fix long case blocks not split into multiple lines

rdrll opened the pull request:
### Description

Fixes #3850 .

Previously Black doesn't recognize `case` block as one of the syntax blocks need to be wrapped with invisible parentheses, and hence will also not be splitting the `case` blocks into multiple lines even when it's too long. 

Now with (invisible) parentheses correctly wrapped around `case` blocks, the lines are able to be splitted into multiple lines when needed, this also allows the trailing comma rules to be enabled for `case` blocks. Unit test `pattern_matching_long.py` and ``pattern_matching_trailing_comma.py`` has been added to verify these modifications.

### A bit of discussion on one side effect

There's one side effect of this modification brings (and it's the reason why some of the old unit test have been modified).

#### Some background

Python 3.10 and later support matching the final case with an undefined pattern (e.g. `_`) as a default when other cases cannot match, this allow code like all following examples to execute and print the "x" without error:

```
# E.g.1 the usual way for default case
x = 1
match x:
    case 10:
        print(1)
    case _:
        print("x")
```

```
# E.g.2 the unusual way 1
x = 1
match x:
    case 10:
        print(1)
    case xxxxxx:
        print("x")
```

```
# E.g.3 the unusual way 2
x = 1
match x:
    case 10:
        print(1)
    case case: # note that here the second "case" is an undefined pattern instead of a Python keyword
        print("x")
```

#### The side effect

After enabling parentheses to be wrapped around `case` blocks (hence also line breaking), the output of this particular corner case in our unit test will be different:
```
match x:
    case case:
        pass
```

Previously, `case` blocks were not considered for wrapping with parentheses, but now Black treats the second occurrence of "case" as an actual `case` block, so **an additional space** will be added between this occurrence and the colon. The underlying representation Black used for the second line is: "case (case ()):").

Hence, the output is now:
```
match x:
    case case :
        pass
```

On the other hand, while the consecutive `case` block now facing this additional space, the consecutive `match` block remains unchanged for unit tests like the one below, since the `match` block have no need of line breaking and was not using invisible parentheses.

```
match match:
    case x:
        pass
```


### Summary

We're currently able to break long `case` blocks into multiple lines and enabled trailing comma rules for it. The previously discussed side effects only affected this particular `case case:` scenario, and it **doesn't** affect any other of the complex nested pattern matching for `case` block under "pattern_matching_complex.py", "pattern_matching_extras.py", and others. 

I am uncertain if it is worthwhile or preferable to add a very special patch for this particular `case case:` scenario. Given that it is an odd and unlikely real-world use case, and considering that the code is cleaner without accommodating this scenario, I lean towards not addressing it specifically. Even with Python 3.10+ allowing to treat the second "case" as an undefined pattern instead of a keyword, I would doubt developers will intentionally use a Python keyword as a pattern/variable.

I removed multiple unit tests that expected `case case:` to be unchanged, and added one unit test expects "case case:" to become "case case :".

Please let me know if changes are needed.

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [ ] Add new / update outdated documentation? not necessary

JelleZijlstra commented:
Thanks, I think we need to fix the `case case :` thing. I would classify the extra space there as a bug. If our parser sees this as nested case blocks, that's an implementation detail that shouldn't leak to user-visible behavior.

JelleZijlstra commented on CHANGES.md:
```diff
@@ -12,6 +12,8 @@
 - Fix crash when whitespace followed a backslash before newline in a docstring (#4008)
 
 - Fix crash on formatting code like `await (a ** b)` (#3994)
+- Fix a bug that long `case` blocks will not be split into multiple lines, also enabled
```
This needs to be changed only in the preview style.

rdrll replied:
I have now added a fix for the `case case:` issue. Since there will be no user-visible format changes, does this still need to be moved to a preview style?

JelleZijlstra replied:
Your change changes formatting (in that lines that previously weren't split now are), so it should go into the preview style. (You should add a new entry to the Preview enum and make your change apply only if that preview feature is on.)

rdrll replied:
Ohh gotcha, thanks! I've now moved it under preview style!

rdrll commented:
Thanks for the feedback, I've implemented a fix for the `case case:` issue. 

Now our parser won't attempt to insert the additional space. The unit tests were also modified (reverted) correspondingly to expect no change for this case. 

This additional fix caused flake8 to flag the `normalize_invisible_parens()` function under `linegen.py` to be C901 too complex (>10 branching logic). Since this function have been the place where we deal with different special cases, I've added an `# noqa: C901` comment to suppress the C901 on this function.

Please let me know if changes are needed.

JelleZijlstra commented on src/black/linegen.py:
```diff
@@ -1260,6 +1260,13 @@ def normalize_invisible_parens(
 
+        # Fixes a bug where invisible parens are not properly wrapped around
+        # case blocks.
+        if isinstance(child, Node) and child.type == syms.case_block:
```
You also need to gate these code changes on the preview style. Look for similar checks like `if Preview.fix_power_op_line_length in mode`. In the test you'll want to add `# flags: --preview`.

rdrll replied:
Thanks for the guidance! I've made the modifications!

JelleZijlstra commented on CHANGES.md:
```diff
@@ -22,6 +22,8 @@
   indented less (#3964)
 - Multiline list and dict unpacking as the sole argument to a function is now also
   indented less (#3992)
+- Fix a bug that long `case` blocks will not be split into multiple lines, while now
+  also enabled general trailing comma rules on `case` blocks (#4024)
```
```suggestion
- Fix a bug where long `case` blocks were not split into multiple lines. Also
  enable general trailing comma rules on `case` blocks (#4024)
```

JelleZijlstra commented:
Not sure what's going on with the diff-shades job; it passed before but started failing after I merged in main. Possibly it's hanging, as I don't see useful output.

rdrll commented:
Yes, this also occured to me earlier [once](https://github.com/psf/black/actions/runs/6763457381) where the diff-shades job failed after running for 1h+, but then it succeeded upon re-running with no actual code change. Maybe I could make some trivial change and see if it can success this time.
